File: apps/windows/src/LexiconBar.Core/BurstDetector.cs
namespace LexiconBar;

/// <summary>
/// A span of text that arrived in the focused field in one "burst": what a
/// dictation tool inserted, as opposed to what the user typed key by key.
/// </summary>
/// <param name="Span">Where the burst sits inside <paramref name="FullText"/>, in UTF-16 units.</param>
/// <param name="Text">The burst itself.</param>
/// <param name="FullText">
/// The whole field value the burst was cut from, so the engine can refuse to
/// rewrite when the field moved on before the API answered.
/// </param>
public sealed record Burst(TextSpan Span, string Text, string FullText);

/// <summary>What <see cref="BurstDetector.Settle"/> decided.</summary>
public abstract record SettleOutcome
{
    private SettleOutcome() { }

    /// <summary>Nothing pending, or not quiet long enough yet.</summary>
    public sealed record Waiting : SettleOutcome
    {
        public static readonly Waiting Instance = new();
    }

    /// <summary>
    /// Dictation is still streaming in: word-sized insertions have been
    /// arriving and the field has not been quiet long enough to call the
    /// phrase finished. Nothing has been decided and the baseline has not
    /// moved — ask again shortly.
    /// </summary>
    public sealed record Coalescing : SettleOutcome
    {
        public static readonly Coalescing Instance = new();
    }

    /// <summary>Pending changes were examined and dismissed; the baseline advanced.</summary>
    public sealed record Skipped(string Reason) : SettleOutcome;

    /// <summary>A burst to send to the API.</summary>
    public sealed record Fired(Burst Burst) : SettleOutcome;
}

/// <summary>
/// Turns a stream of field-value snapshots into bursts. Pure and clock-free:
/// the caller feeds <see cref="Record"/> on every value change and asks
/// <see cref="Settle"/> once it has been quiet for a while.
///
/// One detector per focused element. <see cref="LastStable"/> is the text the
/// field held after the previous decision (fired, skipped or the engine's own
/// rewrite), so the engine's rewrite never re-triggers and typed text that was
/// skipped once is not counted again.
///
/// A burst fires when the field has been quiet for <c>SettleMs</c> and the text
/// inserted since <see cref="LastStable"/>:
/// <list type="bullet">
/// <item>contains at least <c>MinWords</c> words, or arrived in at most
///   <c>FewEvents</c> change events with at least <c>FewEventsMinChars</c> units; and</item>
/// <item>contains at least one change event that inserted <c>MinChunk</c> units
///   at once (dictation lands as words or phrases; typing lands one unit at a
///   time, so key-by-key typing never fires); and</item>
/// <item>is not multi-paragraph (a newline followed by more text) and the field
///   is not longer than <c>MaxFieldLength</c>.</item>
/// </list>
///
/// A direct port of <c>BurstDetector.swift</c>. The four rules that look
/// over-engineered — chunk size rather than rate, coalescing a streamed run,
/// anchoring the window to the caret, and refusing a stale snapshot — each
/// came from a real text-corruption bug on macOS. Do not simplify them.
/// </summary>
public sealed class BurstDetector
{
    /// <param name="FewEvents">Bursts that arrived in this many events or fewer fire at <c>FewEventsMinChars</c> units even if short on words.</param>
    /// <param name="MinChunk">At least one event must insert this many units at once.</param>
    /// <param name="RunQuietMs">
    /// Quiet needed to call a *streamed* run finished, as opposed to a single
    /// insertion. Speech pauses are routinely longer than <c>SettleMs</c>, so a
    /// streamed phrase is only finished after a silence longer than a pause
    /// between words.
    /// </param>
    /// <param name="MaxRunMs">
    /// Hard cap on how long one run may go on coalescing, so continuous
    /// dictation is still corrected periodically instead of never.
    /// </param>
    public sealed record Config(
        int SettleMs = 700,
        int MinWords = 3,
        int FewEvents = 3,
        int FewEventsMinChars = 12,
        int MinChunk = 4,
        int MaxFieldLength = 20_000,
        int RunQuietMs = 1500,
        int MaxRunMs = 12_000)
    {
        public static readonly Config Default = new();
    }

    public Config Settings { get; set; }

    /// <summary>The text the field held after the last decision.</summary>
    public string LastStable { get; private set; }

    /// <summary>The most recent snapshot fed to <see cref="Record"/>.</summary>
    public string Latest { get; private set; }

    private double? _lastChangeAt;

    /// <summary>
    /// When the first change since <see cref="LastStable"/> arrived, so a
    /// coalescing run can be capped by age as well as by silence.
    /// </summary>
    private double? _runStartedAt;

    private int _eventCount;
    private int _maxChunk;

    public BurstDetector(string initialText, Config? config = null)
    {
        Settings = config ?? Config.Default;
        LastStable = initialText;
        Latest = initialText;
    }

    public bool HasPendingChanges => _lastChangeAt is not null;

    /// <summary>Feed the field's current value. Identical values are ignored.</summary>
    public void Record(string text, double atSeconds)
    {
        if (text == Latest) return;
        TextDiff.Delta step = TextDiff.ComputeDelta(Latest, text);
        Latest = text;
        if (_lastChangeAt is null && _runStartedAt is null)
        {
            _eventCount = 0;
            _maxChunk = 0;
            _runStartedAt = atSeconds;
        }

        _eventCount++;
        _maxChunk = Math.Max(_maxChunk, step.Inserted);
        _lastChangeAt = atSeconds;
        if (text == LastStable)
        {
            // Back where we started (undo, or a rewrite landing): nothing pending.
            ClearPending();
        }
    }

    /// <summary>The engine rewrote the field itself: treat <paramref name="text"/> as settled.</summary>
    public void MarkStable(string text)
    {
        LastStable = text;
        Latest = text;
        ClearPending();
    }

    /// <summary>
    /// Ask whether the pending changes form a burst. Call after <c>SettleMs</c>
    /// of quiet; it is safe to call more often.
    /// </summary>
    /// <param name="caretEnd">
    /// Where the insertion point sits now (UTF-16 offset into the field). It
    /// decides *which* of several textually identical windows was the one that
    /// arrived; without it an ambiguous insertion is skipped rather than
    /// guessed at, because rewriting the wrong window corrupts the text around it.
    /// </param>
    public SettleOutcome Settle(double atSeconds, int? caretEnd = null)
    {
        if (_lastChangeAt is not double lastChangeAt) return SettleOutcome.Waiting.Instance;
        double quietMs = (atSeconds - lastChangeAt) * 1000;
        if (quietMs < Settings.SettleMs - 0.5) return SettleOutcome.Waiting.Instance;

        string text = Latest;
        TextDiff.Delta delta = TextDiff.ComputeDelta(LastStable, text, caretEnd);
        int events = _eventCount;
        int chunk = _maxChunk;
        double runAgeMs = _runStartedAt is double started ? (atSeconds - started) * 1000 : 0;
        Verdict verdict = Evaluate(text, delta, events, chunk);

        // Streamed dictation (Windows Voice Access, Wispr Flow, an app's own
        // microphone button) does not arrive as one insertion. It arrives a
        // word at a time, and the gaps between spoken words are routinely
        // longer than SettleMs — so deciding at SettleMs chops the phrase into
        // single words, each of which is "too short" and is dropped. The whole
        // sentence then goes uncorrected, which looks exactly like the feature
        // not working.
        //
        // So: once a run has produced at least one word-sized insertion, keep
        // it pending until the field has been quiet for RunQuietMs — longer
        // than a pause between words, shorter than the end of a sentence —
        // rather than deciding at the first gap. LastStable does not move, so
        // the words accumulate into one burst and are corrected together.
        //
        // Two things keep this from swallowing ordinary typing and pastes:
        // key-by-key typing never produces a MinChunk-sized insertion, so it is
        // never a run and is still refused with "typed, not dictated"; and a
        // single insertion that already reads as a burst is a paste with
        // nothing to wait for, so it fires at SettleMs as before.
        //
        // Only a verdict that more speech could actually change is worth
        // waiting on. Everything else — multi-paragraph, field too long,
        // key-by-key typing, an insertion point we cannot locate — is refused
        // now, because silence will not make it true.
        bool canGrow = verdict switch
        {
            Verdict.Fired => events > 1,
            Verdict.TooShort => true,
            _ => false,
        };
        if (canGrow && quietMs < Settings.RunQuietMs && runAgeMs < Settings.MaxRunMs)
        {
            return SettleOutcome.Coalescing.Instance;
        }

        // Whatever we decide, this is the new baseline.
        LastStable = text;
        ClearPending();
        return verdict switch
        {
            Verdict.Fired fired => new SettleOutcome.Fired(fired.Burst),
            Verdict.TooShort tooShort => new SettleOutcome.Skipped(tooShort.Why),
            Verdict.Refused refused => new SettleOutcome.Skipped(refused.Why),
            _ => SettleOutcome.Waiting.Instance,
        };
    }

    /// <summary>What <see cref="Settle"/> decided, before it decides whether to act on it.</summary>
    private abstract record Verdict
    {
        private Verdict() { }

        public sealed record Fired(Burst Burst) : Verdict;

        /// <summary>Dictation-shaped, but not enough of it yet. More speech changes this.</summary>
        public sealed record TooShort(string Why) : Verdict;

        /// <summary>No amount of waiting changes this.</summary>
        public sealed record Refused(string Why) : Verdict;
    }

    /// <summary>
    /// The burst rules, with no side effects, so <see cref="Settle"/> can look
    /// at the answer and still leave the run pending.
    /// </summary>
    private Verdict Evaluate(string text, TextDiff.Delta delta, int events, int chunk)
    {
        if (delta.Inserted <= 0) return new Verdict.Refused("nothing inserted");
        if (text.Length > Settings.MaxFieldLength)
        {
            return new Verdict.Refused($"field longer than {Settings.MaxFieldLength}");
        }

        if (TextDiff.Substring(text, delta.InsertedSpan) is not string inserted)
        {
            return new Verdict.Refused("range out of bounds");
        }

        // A trailing newline (dictation ending with "new line", or Return)
        // stays out of the burst so the rewrite never has to re-insert one.
        TextSpan span = delta.InsertedSpan;
        while (inserted.Length > 0 && (inserted[^1] == '\n' || inserted[^1] == '\r'))
        {
            inserted = inserted[..^1];
            span = span with { Length = span.Length - 1 };
        }

        if (span.Length <= 0) return new Verdict.Refused("nothing inserted");
        if (chunk < Settings.MinChunk) return new Verdict.Refused("typed, not dictated");

        int words = TextDiff.WordCount(inserted);
        bool enough = words >= Settings.MinWords
                      || (events <= Settings.FewEvents && inserted.Length >= Settings.FewEventsMinChars);
        if (!enough) return new Verdict.TooShort($"too short ({words} words, {inserted.Length} units)");
        if (IsMultiParagraph(inserted)) return new Verdict.Refused("multi-paragraph");

        // Last, because it is the only check whose answer is "this really does
        // look like a burst, but we cannot say where it went". Rewriting a
        // window we only guessed at splices the correction into the middle of
        // the user's own words, so there is nothing to do but leave it alone.
        if (delta.IsAmbiguous)
        {
            return new Verdict.Refused(
                $"insertion point is ambiguous ({delta.SlideLeft} left, {delta.SlideRight} right)");
        }

        return new Verdict.Fired(new Burst(span, inserted, text));
    }

    /// <summary>A newline followed by more (non-whitespace) text. A trailing newline is fine.</summary>
    public static bool IsMultiParagraph(string text)
    {
        for (int i = 0; i < text.Length; i++)
        {
            if (!TextDiff.IsNewline(text[i])) continue;
            for (int j = i + 1; j < text.Length; j++)
            {
                if (!char.IsWhiteSpace(text[j])) return true;
            }
            return false;
        }
        return false;
    }

    private void ClearPending()
    {
        _lastChangeAt = null;
        _runStartedAt = null;
        _eventCount = 0;
        _maxChunk = 0;
    }
}

File: apps/macos/LexiconBar/Sources/LexiconBarKit/BurstDetector.swift
import Foundation

/// A span of text that arrived in the focused field in one "burst": what a
/// dictation tool inserted, as opposed to what the user typed key by key.
/// `range` is in UTF-16 units of `fullText` (the unit Accessibility ranges use).
public struct Burst: Equatable, Sendable {
    public let range: NSRange
    public let text: String
    /// The whole field value the burst was cut from, so the engine can refuse
    /// to rewrite when the field moved on before the API answered.
    public let fullText: String

    public init(range: NSRange, text: String, fullText: String) {
        self.range = range
        self.text = text
        self.fullText = fullText
    }
}

/// UTF-16 diff helpers shared by the detector and the splice math.
public enum TextDiff {
    /// The span of `new` that is not shared with `old`: the common prefix and
    /// suffix (in UTF-16 units, never overlapping) are stripped and what
    /// remains is the insertion. `removed` is the corresponding span of `old`.
    ///
    /// A prefix/suffix diff cannot always say *where* an insertion happened.
    /// Pasting "ping ashler …" in front of a field that already reads
    /// "ping Ashlr.AI …" shares the leading "ping ", so the stripped window
    /// starts five units late and runs five units into the old text — a span
    /// that is textually consistent but is not what was inserted. Rewriting
    /// that window splices the correction into the middle of a word and leaves
    /// the rest of the burst untouched. `slideLeft` / `slideRight` measure how
    /// far the window could move and still produce `new`; `anchored` says the
    /// caret pinned it, so the position is known rather than guessed.
    public struct Delta: Equatable, Sendable {
        public let location: Int
        public let inserted: Int
        public let removed: Int
        /// Play in the insertion point, in UTF-16 units. Zero for a replacement.
        public let slideLeft: Int
        public let slideRight: Int
        /// The caret picked `location` out of the possible positions.
        public let anchored: Bool

        public init(location: Int, inserted: Int, removed: Int,
                    slideLeft: Int = 0, slideRight: Int = 0, anchored: Bool = false) {
            self.location = location
            self.inserted = inserted
            self.removed = removed
            self.slideLeft = slideLeft
            self.slideRight = slideRight
            self.anchored = anchored
        }

        public var insertedRange: NSRange { NSRange(location: location, length: inserted) }

        /// The insertion could have happened somewhere else and nothing pinned
        /// it down. The caller must not rewrite a window it only guessed at.
        public var isAmbiguous: Bool { !anchored && (slideLeft > 0 || slideRight > 0) }
    }

    /// `caretEnd` is the UTF-16 offset the insertion point sits at now. After
    /// a paste or a dictation insert it is the end of what arrived, which is
    /// what resolves an ambiguous window. Pass nil when it is unknown.
    public static func delta(from old: String, to new: String, caretEnd: Int? = nil) -> Delta {
        let a = Array(old.utf16)
        let b = Array(new.utf16)
        var prefix = 0
        let maxPrefix = min(a.count, b.count)
        while prefix < maxPrefix, a[prefix] == b[prefix] { prefix += 1 }
        var suffix = 0
        let maxSuffix = maxPrefix - prefix
        while suffix < maxSuffix, a[a.count - 1 - suffix] == b[b.count - 1 - suffix] { suffix += 1 }
        let location = prefix
        let inserted = b.count - prefix - suffix
        let removed = a.count - prefix - suffix
        guard removed == 0, inserted > 0 else {
            return Delta(location: location, inserted: inserted, removed: removed)
        }

        // How far the window can slide while `new` stays the same: one unit to
        // the left whenever the unit before the window equals the last unit in
        // it, and one to the right whenever the unit after it equals the first.
        var left = 0
        while location - left > 0, b[location - left - 1] == b[location - left + inserted - 1] { left += 1 }
        var right = 0
        while location + inserted + right < b.count, b[location + right] == b[location + inserted + right] { right += 1 }

        if let caretEnd, caretEnd - inserted >= location - left, caretEnd - inserted <= location + right {
            let anchoredLocation = caretEnd - inserted
            let shift = anchoredLocation - location
            return Delta(location: anchoredLocation, inserted: inserted, removed: 0,
                         slideLeft: left + shift, slideRight: right - shift, anchored: true)
        }
        return Delta(location: location, inserted: inserted, removed: 0, slideLeft: left, slideRight: right)
    }

    /// Substring of `text` by UTF-16 range; nil when the range does not fit.
    public static func substring(_ text: String, utf16 range: NSRange) -> String? {
        guard range.location >= 0, range.length >= 0, range.location + range.length <= text.utf16.count else { return nil }
        let utf16 = text.utf16
        let start = utf16.index(utf16.startIndex, offsetBy: range.location)
        let end = utf16.index(start, offsetBy: range.length)
        return String(utf16[start..<end])
    }

    /// Replaces the UTF-16 `range` of `text` with `replacement`; nil when the
    /// range does not fit (the caller must then leave the field alone).
    public static func splice(_ text: String, utf16 range: NSRange, with replacement: String) -> String? {
        guard range.location >= 0, range.length >= 0, range.location + range.length <= text.utf16.count else { return nil }
        let utf16 = text.utf16
        let start = utf16.index(utf16.startIndex, offsetBy: range.location)
        let end = utf16.index(start, offsetBy: range.length)
        var units = Array(utf16[..<start])
        units.append(contentsOf: replacement.utf16)
        units.append(contentsOf: utf16[end...])
        return String(decoding: units, as: UTF16.self)
    }

    public static func wordCount(_ text: String) -> Int {
        text.split(whereSeparator: { $0.isWhitespace || $0.isNewline }).count
    }
}

/// Turns a stream of field-value snapshots into bursts. Pure and clock-free:
/// the caller feeds `record(text:at:)` on every value change and asks
/// `settle(at:)` once it has been quiet for a while.
///
/// One detector per focused element. `lastStable` is the text the field held
/// after the previous decision (fired, skipped or the engine's own rewrite),
/// so the engine's rewrite never re-triggers and typed text that was skipped
/// once is not counted again.
///
/// A burst fires when the field has been quiet for `settleMs` and the text
/// inserted since `lastStable`:
///  - contains at least `minWords` words, or arrived in at most three change
///    events with at least 12 UTF-16 units; and
///  - contains at least one change event that inserted `minChunk` units at
///    once (dictation lands as words or phrases; typing lands one unit at a
///    time, so key-by-key typing never fires); and
///  - is not multi-paragraph (a newline followed by more text) and the field
///    is not longer than `maxFieldLength`.
public final class BurstDetector {
    public struct Config: Equatable, Sendable {
        public var settleMs: Int
        public var minWords: Int
        /// Bursts that arrived in <= `fewEvents` events fire at `fewEventsMinChars` units even if short on words.
        public var fewEvents: Int
        public var fewEventsMinChars: Int
        /// At least one event must insert this many units at once.
        public var minChunk: Int
        public var maxFieldLength: Int
        /// Quiet needed to call a *streamed* run finished, as opposed to a
        /// single insertion. Speech pauses are routinely longer than
        /// `settleMs`, so a streamed phrase is only finished after a silence
        /// longer than a pause between words.
        public var runQuietMs: Int
        /// Hard cap on how long one run may go on coalescing, so continuous
        /// dictation is still corrected periodically instead of never.
        public var maxRunMs: Int

        public init(settleMs: Int = 700, minWords: Int = 3, fewEvents: Int = 3, fewEventsMinChars: Int = 12,
                    minChunk: Int = 4, maxFieldLength: Int = 20_000,
                    runQuietMs: Int = 1500, maxRunMs: Int = 12_000) {
            self.settleMs = settleMs
            self.minWords = minWords
            self.fewEvents = fewEvents
            self.fewEventsMinChars = fewEventsMinChars
            self.minChunk = minChunk
            self.maxFieldLength = maxFieldLength
            self.runQuietMs = runQuietMs
            self.maxRunMs = maxRunMs
        }

        public static let `default` = Config()
    }

    public enum Outcome: Equatable, Sendable {
        /// Nothing pending, or not quiet long enough yet.
        case waiting
        /// Dictation is still streaming in: word-sized insertions have been
        /// arriving and the field has not been quiet long enough to call the
        /// phrase finished. Nothing has been decided and `lastStable` has not
        /// moved — ask again shortly.
        case coalescing
        /// Pending changes were examined and dismissed; `lastStable` advanced.
        case skipped(String)
        case burst(Burst)
    }

    public var config: Config
    public private(set) var lastStable: String
    public private(set) var latest: String
    private var lastChangeAt: TimeInterval?
    /// When the first change since `lastStable` arrived, so a coalescing run
    /// can be capped by age as well as by silence.
    private var runStartedAt: TimeInterval?
    private var eventCount = 0
    private var maxChunk = 0

    public init(initialText: String, config: Config = .default) {
        self.config = config
        self.lastStable = initialText
        self.latest = initialText
    }

    public var hasPendingChanges: Bool { lastChangeAt != nil }

    /// Feed the field's current value. Identical values are ignored.
    public func record(text: String, at time: TimeInterval) {
        guard text != latest else { return }
        let step = TextDiff.delta(from: latest, to: text)
        latest = text
        if lastChangeAt == nil, runStartedAt == nil {
            eventCount = 0
            maxChunk = 0
            runStartedAt = time
        }
        eventCount += 1
        maxChunk = max(maxChunk, step.inserted)
        lastChangeAt = time
        if text == lastStable {
            // Back where we started (undo, or a rewrite landing): nothing pending.
            clearPending()
        }
    }

    /// The engine rewrote the field itself: treat `text` as settled.
    public func markStable(_ text: String) {
        lastStable = text
        latest = text
        clearPending()
    }

    /// Ask whether the pending changes form a burst. Call after `settleMs`
    /// of quiet; it is safe to call more often.
    ///
    /// `caretEnd` is where the insertion point sits now (UTF-16 offset into
    /// the field). It decides *which* of several textually identical windows
    /// was the one that arrived; without it an ambiguous insertion is skipped
    /// rather than guessed at, because rewriting the wrong window corrupts the
    /// text around it.
    public func settle(at time: TimeInterval, caretEnd: Int? = nil) -> Outcome {
        guard let lastChangeAt else { return .waiting }
        let quietMs = (time - lastChangeAt) * 1000
        guard quietMs >= Double(config.settleMs) - 0.5 else { return .waiting }
        let text = latest
        let delta = TextDiff.delta(from: lastStable, to: text, caretEnd: caretEnd)
        let events = eventCount
        let chunk = maxChunk
        let runAgeMs = runStartedAt.map { (time - $0) * 1000 } ?? 0
        let verdict = evaluate(text: text, delta: delta, events: events, chunk: chunk)

        // Streamed dictation (macOS dictation, Wispr Flow's streaming mode,
        // an app's own microphone button) does not arrive as one insertion.
        // It arrives a word at a time, and the gaps between spoken words are
        // routinely longer than `settleMs` — so deciding at `settleMs` chops
        // the phrase into single words, each of which is "too short" and is
        // dropped. The whole sentence then goes uncorrected, which looks
        // exactly like the feature not working.
        //
        // So: once a run has produced at least one word-sized insertion, keep
        // it pending until the field has been quiet for `runQuietMs` — longer
        // than a pause between words, shorter than the end of a sentence —
        // rather than deciding at the first gap. `lastStable` does not move,
        // so the words accumulate into one burst and are corrected together.
        //
        // Two things keep this from swallowing ordinary typing and pastes:
        // key-by-key typing never produces a `minChunk`-sized insertion, so
        // it is never a run and is still refused with "typed, not dictated";
        // and a single insertion that already reads as a burst is a paste
        // with nothing to wait for, so it fires at `settleMs` as before.
        // Only a verdict that more speech could actually change is worth
        // waiting on. "Too short" can grow into a burst; a burst that arrived
        // as several insertions may still have words coming. A single
        // insertion that already reads as a burst is a paste, with nothing to
        // wait for, so it still fires at `settleMs`. Everything else —
        // multi-paragraph, field too long, key-by-key typing, an insertion
        // point we cannot locate — is refused now, because silence will not
        // make it true.
        let canGrow: Bool
        switch verdict {
        case .burst: canGrow = events > 1
        case .tooShort: canGrow = true
        case .refused: canGrow = false
        }
        if canGrow, quietMs < Double(config.runQuietMs), runAgeMs < Double(config.maxRunMs) {
            return .coalescing
        }

        // Whatever we decide, this is the new baseline.
        lastStable = text
        clearPending()
        switch verdict {
        case .burst(let burst): return .burst(burst)
        case .tooShort(let why), .refused(let why): return .skipped(why)
        }
    }

    /// What `settle` decided, before it decides whether to act on it.
    private enum Verdict {
        case burst(Burst)
        /// Dictation-shaped, but not enough of it yet. More speech changes this.
        case tooShort(String)
        /// No amount of waiting changes this.
        case refused(String)
    }

    /// The burst rules, with no side effects, so `settle` can look at the
    /// answer and still leave the run pending.
    private func evaluate(text: String, delta: TextDiff.Delta, events: Int, chunk: Int) -> Verdict {
        guard delta.inserted > 0 else { return .refused("nothing inserted") }
        guard text.utf16.count <= config.maxFieldLength else { return .refused("field longer than \(config.maxFieldLength)") }
        guard var inserted = TextDiff.substring(text, utf16: delta.insertedRange) else { return .refused("range out of bounds") }
        // A trailing newline (dictation ending with "new line", or Return) stays
        // out of the burst so the rewrite never has to re-insert one.
        var range = delta.insertedRange
        while let last = inserted.unicodeScalars.last, last == "\n" || last == "\r" {
            inserted.unicodeScalars.removeLast()
            range.length -= 1
        }
        guard range.length > 0 else { return .refused("nothing inserted") }
        guard chunk >= config.minChunk else { return .refused("typed, not dictated") }
        let words = TextDiff.wordCount(inserted)
        let enough = words >= config.minWords || (events <= config.fewEvents && inserted.utf16.count >= config.fewEventsMinChars)
        guard enough else { return .tooShort("too short (\(words) words, \(inserted.utf16.count) units)") }
        guard !BurstDetector.isMultiParagraph(inserted) else { return .refused("multi-paragraph") }
        // Last, because it is the only check whose answer is "this really does
        // look like a burst, but we cannot say where it went". Rewriting a
        // window we only guessed at splices the correction into the middle of
        // the user's own words, so there is nothing to do but leave it alone.
        guard !delta.isAmbiguous else {
            return .refused("insertion point is ambiguous (\(delta.slideLeft) left, \(delta.slideRight) right)")
        }
        return .burst(Burst(range: range, text: inserted, fullText: text))
    }

    /// A newline followed by more (non-whitespace) text. A trailing newline is fine.
    public static func isMultiParagraph(_ text: String) -> Bool {
        guard let newline = text.firstIndex(where: { $0.isNewline }) else { return false }
        let rest = text[text.index(after: newline)...]
        return rest.contains(where: { !$0.isWhitespace })
    }

    private func clearPending() {
        lastChangeAt = nil
        runStartedAt = nil
        eventCount = 0
        maxChunk = 0
    }
}
